sim/common/cata/dragonwrath.go

```diff
@@ -0,0 +1,280 @@
+	dotSpell.BonusSpellPower = result.Damage
+	dotSpell.Cast(sim, result.Target)
+}
+
+func defaultSpellHandler(sim *core.Simulation, spell *core.Spell, result *core.SpellResult) {
+	copySpell := spell.Unit.GetSpell(spell.WithTag(71086))
+	if copySpell == nil {
+		copySpell = spell.Unit.RegisterSpell(GetDRTSpellConfig(spell))
+	}
+
+	sim.AddPendingAction(
+		&core.PendingAction{
+			NextActionAt: sim.CurrentTime + time.Millisecond*200, // add slight delay
+			Priority:     core.ActionPriorityAuto,
+			OnAction: func(sim *core.Simulation) {
+				copySpell.Cast(sim, result.Target)
+			},
+		},
+	)
+}
+
+func GetDRTSpellConfig(spell *core.Spell) core.SpellConfig {
+	baseConfig := core.SpellConfig{
+		ActionID:     spell.WithTag(71086),
+		SpellSchool:  spell.SpellSchool,
+		ProcMask:     core.ProcMaskEmpty,
+		ApplyEffects: spell.ApplyEffects,
+		ManaCost:     core.ManaCostOptions{},
+		Cast: core.CastConfig{
+			DefaultCast: core.Cast{
+				NonEmpty: true,
+			},
+		},
+		CritMultiplier:           spell.Unit.Env.Raid.GetPlayerFromUnit(spell.Unit).GetCharacter().DefaultSpellCritMultiplier(),
+		DamageMultiplier:         1,
+		DamageMultiplierAdditive: 1,
+		MissileSpeed:             spell.MissileSpeed,
+		ClassSpellMask:           spell.ClassSpellMask,
+		BonusCoefficient:         spell.BonusCoefficient,
+		Flags:                    spell.Flags &^ core.SpellFlagAPL,
+	}
+
+	// copy BonusCoefficient as spells might have spell mods applied to them
+	// we dont know the original in the base spell config so they might have been applied twice
+	baseConfig.BonusCoefficient = spell.BonusCoefficient
+	return baseConfig
+}
+
+func init() {
+	core.NewItemEffect(71086, func(a core.Agent) {
+		unit := &a.GetCharacter().Unit
+		registerSpells(unit)
+
+		lastTimestamp := time.Duration(0)
+		spellList := map[int32]bool{}
+		core.MakePermanent(unit.RegisterAura(core.Aura{
+			ActionID: core.ActionID{ItemID: 71086},
+			Label:    "Dragonwrath, Tarecgosa's Rest - Handler",
+			OnSpellHitDealt: func(aura *core.Aura, sim *core.Simulation, spell *core.Spell, result *core.SpellResult) {
+
+				// Handle direct damage only and make sure we're not proccing of our own spell
+				if !result.Landed() || spell.ActionID.Tag == 71086 {
+					return
+				}
+
+				// there are many apply effects with 0 damage right now for spells that actually have no impact damage
+				// we do not want replicate those
+				if result.PreOutcomeDamage == 0 {
+					return
+				}
+
+				if !spell.ProcMask.Matches(core.ProcMaskSpellOrProc) {
+					return
+				}
+
+				config := classConfig[a.GetCharacter().Spec]
+				if config == nil {
+					// TODO: HANDLE BETTER
+					panic("DTR not supported for this spec yet")
+				}
+
+				if val, ok := config.spellConfig[spell.SpellID]; ok && val&supressImpact > 0 {
+					return
+				}
+
+				// make sure the same spell impact can only trigger once per timestamp (AoE Impact spells like Arcane Explosion or Mind Sear)
```

Comment:
This might prevent multiple Chain Lightning Overloads from being duplicated

Reply:
You might be right. Trigger/Proc systems are always the hardest to get right even when I was working on p-servers as they heavily depend on how blizz actually implemented stuff.

I suppose we have the following:
* `TARGET_UNIT_TARGET_ENEMY` - proc per target even chain targets
* `TARGET_UNIT_DEST_AREA_ENEMY` and `TARGET_UNIT_SRC_AREA_ENEMY` - per cast no matter how many targets
  * Usually those are spells like  concecrate, blizzard, arcane explosion or mind sear.

Obviously we don't have access to the target enums for the spells but I guess we need to be able to configure those spells in the class setup. I'll take a look to provide a clean API.

Reply:
I more and more wish we had a differenitation between spells and casts i guess xD